conduit/pipeline_test.go

```diff
@@ -0,0 +1,223 @@
+	assert.Equal(t, pCfg.Processors[0].Name, "noop")
+	assert.Equal(t, pCfg.Processors[0].Config["catchpoint"], "7560000#3OUX3TLXZNOK6YJXGETKRRV2MHMILF5CCIVZUOJCT6SLY5H2WWTQ")
+	assert.Equal(t, pCfg.Exporter.Name, "noop")
+	assert.Equal(t, pCfg.Exporter.Config["connectionstring"], "")
+
+	// "" for dir will use os.TempDir()
+	// invalidDataDir has no auto load file
+	invalidDataDir, err := ioutil.TempDir("", "conduit_data_dir")
+	assert.Nil(t, err)
+	defer os.RemoveAll(invalidDataDir)
+
+	cfgBad := &Config{ConduitDataDir: invalidDataDir}
+	_, err = MakePipelineConfig(l, cfgBad)
+	assert.Equal(t, err,
+		fmt.Errorf("MakePipelineConfig(): could not find %s in data directory (%s)", autoLoadParameterConfigName, cfgBad.ConduitDataDir))
+
+}
+
+// a unique block data to validate with tests
+var uniqueBlockData = data.BlockData{
+	BlockHeader: bookkeeping.BlockHeader{
+		Round: 1337,
+	},
+}
+
+type mockImporter struct {
+	mock.Mock
+	importers.Importer
+}
+
+func (m *mockImporter) GetBlock(rnd uint64) (data.BlockData, error) {
+	m.Called(rnd)
+	return uniqueBlockData, nil
+}
+
+type mockProcessor struct {
+	mock.Mock
+	processors.Processor
+	finalRound basics.Round
+}
+
+func (m *mockProcessor) Process(input data.BlockData) (data.BlockData, error) {
+	m.Called(input)
+	input.BlockHeader.Round++
+	return input, nil
+}
+func (m *mockProcessor) OnComplete(input data.BlockData) error {
+	m.finalRound = input.BlockHeader.Round
+	m.Called(input)
+	return nil
+}
+
+type mockExporter struct {
+	mock.Mock
+	exporters.Exporter
+}
+
+func (m *mockExporter) Receive(exportData data.BlockData) error {
+	m.Called(exportData)
+	return nil
+}
+
+// TestPipelineRun tests that running the pipeline calls the correct functions with mocking
+func TestPipelineRun(t *testing.T) {
+
+	mImporter := mockImporter{}
+	mImporter.On("GetBlock", mock.Anything).Return(uniqueBlockData, nil)
+	mProcessor := mockProcessor{}
+	processorData := uniqueBlockData
+	processorData.BlockHeader.Round++
+	mProcessor.On("Process", mock.Anything).Return(processorData)
+	mProcessor.On("OnComplete", mock.Anything).Return(nil)
+	mExporter := mockExporter{}
+	mExporter.On("Receive", mock.Anything).Return(nil)
+
+	var pImporter importers.Importer = &mImporter
+	var pProcessor processors.Processor = &mProcessor
+	var pExporter exporters.Exporter = &mExporter
+
+	ctx, cf := context.WithCancel(context.Background())
+
+	pImpl := pipelineImpl{
+		ctx:          ctx,
+		cfg:          &PipelineConfig{},
+		logger:       log.New(),
+		initProvider: nil,
+		importer:     &pImporter,
+		processors:   []*processors.Processor{&pProcessor},
+		exporter:     &pExporter,
+		round:        0,
+	}
+
+	go func() {
+		time.Sleep(1 * time.Second)
+		cf()
+	}()
+
+	err := pImpl.RunPipeline()
+	assert.Nil(t, err)
+
+	assert.Equal(t, mProcessor.finalRound, basics.Round(uniqueBlockData.BlockHeader.Round+1))
+
+	mock.AssertExpectationsForObjects(t, &mImporter, &mProcessor, &mExporter)
```

Comment:
Nice! Maybe we add an error for each plugin type as well?

Reply:
Do you mean assert expectations for each object?

Reply:
No I mean add errors to the responses to test error handling as well.

```go
mImporter.On("GetBlock", mock.Anything).Return(uniqueBlockData, fmt.Errorf("foobar"))
err := pImpl.RunPipeline()
assert.Err(t, err)
```